source/common/stats/stats_impl.cc

```diff
@@ -135,35 +135,48 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
   return false;
 }
 
-RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  uint64_t num_bytes_to_allocate = RawStatData::structSize(name.size());
-  RawStatData* data = static_cast<RawStatData*>(::calloc(num_bytes_to_allocate, 1));
-  if (data == nullptr) {
-    throw std::bad_alloc();
-  }
-  data->checkAndInit(name, num_bytes_to_allocate);
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
+  auto data = std::make_unique<HeapStatData>(truncateStatName(name));
   Thread::ReleasableLockGuard lock(mutex_);
-  auto ret = stats_.insert(data);
-  RawStatData* existing_data = *ret.first;
+  auto ret = stats_.insert(data.get());
+  HeapStatData* existing_data = *ret.first;
   lock.release();
 
-  if (!ret.second) {
-    ::free(data);
-    ++existing_data->ref_count_;
-    return existing_data;
-  } else {
-    return data;
+  if (ret.second) {
+    return data.release();
+  }
+  ++existing_data->ref_count_;
+  return existing_data;
+}
+
+void HeapStatDataAllocator::free(HeapStatData& data) {
+  ASSERT(data.ref_count_ > 0);
+  if (--data.ref_count_ > 0) {
+    return;
   }
+
+  {
+    Thread::LockGuard lock(mutex_);
+    size_t key_removed = stats_.erase(&data);
+    ASSERT(key_removed == 1);
+  }
+
+  delete &data;
 }
 
 /**
- * Counter implementation that wraps a RawStatData.
+ * Counter implementation that wraps a StatData. StatData must have data members:
+ *    std::atomic<int64_t> value_;
+ *    std::atomic<int64_t> pending_increment_;
+ *    std::atomic<int16_t> flags_;
+ *    std::atomic<int16_t> ref_count_;
  */
-class CounterImpl : public Counter, public MetricImpl {
+template <class StatData> class CounterImpl : public Counter, public MetricImpl {
```

Comment:
Why do we need to use templates here vs. polymorphism? Would prefer the latter if possible for the usual template reasons (horrible compiler errors being my pet peeve).

Reply:
I'm not following your suggestion.  AFAIK polymorphism is not an alternative to templates, for code-sharing the class bodies. The most direct alternative is macros, and a functional (but less performance and memory-efficient) substitute in this case would be virtual inheritance from a base class. I think templates are the correct solution here.

Reply:
Polymorphism == virtual inheritance for the purpose of my comment :) OK, I'm willing to buy templates for performance reasons since stats increment etc. is actually on the hot path. I think you should add justification at the template declaration sites explicitly calling this out, since we default to "avoid templates unless absolutely necessary" in Envoy, an maybe someone comes along and undoes this if you don't ;)

Reply:
done

Reply:
Just to toss my hat in here - as I mentioned above, I agree with Harvey that the Envoy default is to avoid templates as much as possible.  I agree that this is the most performant solution for sharing code that I can think of.  However, it seems like the majority of the Envoy codebase (even portions in the hot path) use virtual function calls everywhere.  This feels a little bit like premature optimization.  We already do one virtual function call into Counter and Gauge per change since all the functions on those classes are virtual.  Do we have an idea how much an extra virtual call here will affect real-world performance?

I'm not super opinionated on this, and if no one else has issues with using templates this extensively, we can leave it as is, but it seems like this does go against general Envoy convention.

Reply:
It's about 5 instructions for a virtual dispatch. I think it's very likely premature optimization but I didn't look into the change in very deep detail.

Reply:
Added comments in code:
```
//
// Also note that RawStatData needs to live in a shared memory block, and it's
// possible, but not obvious, that a vptr would be usable across processes. In
// any case, RawStatData is currently allocated with calloc() rather than new,
// so the usual C++ compiler assistance for setting up vptrs will not be
// available. This could be resolved with placed new.
```

Reply:
I think the way you would implement this with virtual inheritance is by wrapping a reference to the struct that is stored in shared memory in a virtual class that would be local to a single instance of Envoy.  This is similar to how CounterImpl, etc works today - just more nested. 

Anyway, I'll yield to @mattklein123 and @htuch on this since they are much more aware of what the overall Envoy conventions and perf tradeoffs are.